Source/MLX/MLXArray.swift

```diff
@@ -160,35 +160,30 @@ public final class MLXArray {
     /// let value = array[1].item(Float.self)
     /// ```
     public func item<T: HasDType>(_ type: T.Type) -> T {
-        self.eval()
+        precondition(self.size == 1)
 
-        var array_ctx = self.ctx
-        var free = false
         if type.dtype != self.dtype {
-            array_ctx = mlx_astype(self.ctx, type.dtype.cmlxDtype, StreamOrDevice.default.ctx)
-            mlx_array_eval(array_ctx)
-            free = true
+            return self.asType(type).item(type)
         }
 
-        // can't do it inside the else as it will free at the end of the block
-        defer { if free { mlx_free(array_ctx) } }
+        self.eval()
 
         switch type {
-        case is Bool.Type: return mlx_array_item_bool(array_ctx) as! T
-        case is UInt8.Type: return mlx_array_item_uint8(array_ctx) as! T
-        case is UInt16.Type: return mlx_array_item_uint16(array_ctx) as! T
-        case is UInt32.Type: return mlx_array_item_uint32(array_ctx) as! T
-        case is UInt64.Type: return mlx_array_item_uint64(array_ctx) as! T
-        case is Int8.Type: return mlx_array_item_int8(array_ctx) as! T
-        case is Int16.Type: return mlx_array_item_int16(array_ctx) as! T
-        case is Int32.Type: return mlx_array_item_int32(array_ctx) as! T
-        case is Int64.Type: return mlx_array_item_int64(array_ctx) as! T
-        case is Int.Type: return Int(mlx_array_item_int64(array_ctx)) as! T
+        case is Bool.Type: return mlx_array_item_bool(self.ctx) as! T
+        case is UInt8.Type: return mlx_array_item_uint8(self.ctx) as! T
+        case is UInt16.Type: return mlx_array_item_uint16(self.ctx) as! T
+        case is UInt32.Type: return mlx_array_item_uint32(self.ctx) as! T
+        case is UInt64.Type: return mlx_array_item_uint64(self.ctx) as! T
+        case is Int8.Type: return mlx_array_item_int8(self.ctx) as! T
+        case is Int16.Type: return mlx_array_item_int16(self.ctx) as! T
+        case is Int32.Type: return mlx_array_item_int32(self.ctx) as! T
+        case is Int64.Type: return mlx_array_item_int64(self.ctx) as! T
+        case is Int.Type: return Int(mlx_array_item_int64(self.ctx)) as! T
         #if !arch(x86_64)
-            case is Float16.Type: return mlx_array_item_float16(array_ctx) as! T
+            case is Float16.Type: return mlx_array_item_float16(self.ctx) as! T
         #endif
```

Comment:
Just curious, what happens if the type is `bfloat16`?

Reply:
It will be a fatal error.  This returns a value of type `T` and there is no type that represents `bfloat16`, so we can't return it.  That said, it is not possible to _ask_ for a `bfloat16` so it can't happen -- you could ask for `Float` or `Float16` and it will convert to that.